{"metadata":{"language_info":{"name":"python","version":"3.7.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"import pandas as pd\n\nurl = [\"index_2020.csv\",\"index_2019.csv\",\"index_2018.csv\",\"index_2017.csv\",\"index_2016.csv\",\"index_2015.csv\",\"index_2014.csv\",\"index_2013.csv\",\"index_2012.csv\",\"index_2011.csv\"]\n\ndf_2020 = pd.read_csv(url[0])   #opens all of the index csv files and assigns each one to an individual pandas dataframe\ndf_2019 = pd.read_csv(url[1])\ndf_2018 = pd.read_csv(url[2])\ndf_2017 = pd.read_csv(url[3])\ndf_2016 = pd.read_csv(url[4])\ndf_2015 = pd.read_csv(url[5])\ndf_2014 = pd.read_csv(url[6])\ndf_2013 = pd.read_csv(url[7])\ndf_2012 = pd.read_csv(url[8])\ndf_2011 = pd.read_csv(url[9])\n\ndf_total = pd.concat([df_2020, df_2019, df_2018, df_2017, df_2016, df_2015, df_2014, df_2013, df_2012, df_2011], sort=False)\ndf_total = df_total[[\"OBJECT_ID\"]]   #combines the individual dataframes into one whole and isolates the OBJECT_ID value\n\nxml_links = []\nfor i in df_total[\"OBJECT_ID\"]:   #scans all of the IDs in the list and puts them into link format\n    xml_links.append(\"https://s3.amazonaws.com/irs-form-990/{}_public.xml\".format(i))\n   #this produces a list of approximately 3.2 million AWS links, which all lead to a unique 990 file","metadata":{"trusted":true},"execution_count":1,"outputs":[]},{"cell_type":"code","source":"import xml.etree.ElementTree as et\nimport requests\n\ndf_rows = []\ndf_cols = [\"link (DAF)\"]\n\na = 0\nfor i in xml_links:   #uses a for loop to open each link and read thru the contents of the XML file\n    xml_data = requests.get(xml_links[a]).content\n    xtree = et.fromstring(xml_data)\n    for i in xtree.findall(\".//\"):\n        if 'DonorAdvisedFundInd' in i.tag:   #checks for the <DonorAdvisedFundInd> value\n            if i.text is \"1\":   #if the value is 1 (it is a DAF) the AWS link is printed to console and added to a final dataframe\n                df_rows.append({\"link\":xml_links[a]})\n                print(xml_links[a])\n            elif i.text is \"0\" or \"false\":   #if the value is 0 (not a DAF) the link is passed\n                pass\n    a = a + 1\nout_df = pd.DataFrame(df_rows, columns = df_cols)\nout_df\n   #creates and prints a pandas dataframe that contains links to all the 990 filings that filed true for DAF","metadata":{"trusted":true},"execution_count":2,"outputs":[{"name":"stdout","text":"https://s3.amazonaws.com/irs-form-990/201913199349320656_public.xml\nhttps://s3.amazonaws.com/irs-form-990/201913259349301006_public.xml\n","output_type":"stream"},{"name":"stderr","text":"ERROR:root:Internal Python error in the inspect module.\nBelow is the traceback from this internal error.\n\n","output_type":"stream"},{"name":"stdout","text":"Traceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3331, in run_code\n    exec(code_obj, self.user_global_ns, self.user_ns)\n  File \"<ipython-input-2-a93eb7cd3f2f>\", line 9, in <module>\n    xml_data = requests.get(xml_links[a]).content\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/requests/api.py\", line 76, in get\n    return request('get', url, params=params, **kwargs)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/requests/api.py\", line 61, in request\n    return session.request(method=method, url=url, **kwargs)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/requests/sessions.py\", line 530, in request\n    resp = self.send(prep, **send_kwargs)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/requests/sessions.py\", line 643, in send\n    r = adapter.send(request, **kwargs)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/requests/adapters.py\", line 449, in send\n    timeout=timeout\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/urllib3/connectionpool.py\", line 677, in urlopen\n    chunked=chunked,\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/urllib3/connectionpool.py\", line 426, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/urllib3/connectionpool.py\", line 421, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/srv/conda/envs/notebook/lib/python3.7/http/client.py\", line 1344, in getresponse\n    response.begin()\n  File \"/srv/conda/envs/notebook/lib/python3.7/http/client.py\", line 306, in begin\n    version, status, reason = self._read_status()\n  File \"/srv/conda/envs/notebook/lib/python3.7/http/client.py\", line 267, in _read_status\n    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n  File \"/srv/conda/envs/notebook/lib/python3.7/socket.py\", line 589, in readinto\n    return self._sock.recv_into(b)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/urllib3/contrib/pyopenssl.py\", line 313, in recv_into\n    return self.connection.recv_into(*args, **kwargs)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/OpenSSL/SSL.py\", line 1839, in recv_into\n    result = _lib.SSL_read(self._ssl, buf, nbytes)\nKeyboardInterrupt\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2044, in showtraceback\n    stb = value._render_traceback_()\nAttributeError: 'KeyboardInterrupt' object has no attribute '_render_traceback_'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1148, in get_records\n    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 316, in wrapped\n    return f(*args, **kwargs)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 350, in _fixed_getinnerframes\n    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n  File \"/srv/conda/envs/notebook/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n  File \"/srv/conda/envs/notebook/lib/python3.7/inspect.py\", line 1464, in getframeinfo\n    lines, lnum = findsource(frame)\n  File \"/srv/conda/envs/notebook/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 182, in findsource\n    lines = linecache.getlines(file, globals_dict)\n  File \"/srv/conda/envs/notebook/lib/python3.7/linecache.py\", line 47, in getlines\n    return updatecache(filename, module_globals)\n  File \"/srv/conda/envs/notebook/lib/python3.7/linecache.py\", line 136, in updatecache\n    with tokenize.open(fullname) as fp:\n  File \"/srv/conda/envs/notebook/lib/python3.7/tokenize.py\", line 449, in open\n    encoding, lines = detect_encoding(buffer.readline)\n  File \"/srv/conda/envs/notebook/lib/python3.7/tokenize.py\", line 418, in detect_encoding\n    first = read_or_stop()\n  File \"/srv/conda/envs/notebook/lib/python3.7/tokenize.py\", line 376, in read_or_stop\n    return readline()\nKeyboardInterrupt\n","output_type":"stream"},{"traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"],"ename":"KeyboardInterrupt","evalue":"","output_type":"error"}]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}